src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java

```diff
@@ -599,6 +599,32 @@ public ClassDesc asSymbol() {
             return this.sym = sym;
         }
 
+        @Override
+        public boolean equalsSymbol(ClassDesc symbol) {
+            if (symbol.isPrimitive()) // implicit null check
+                return false;
+
+            var mySym = this.sym;
+            if (mySym != null)
+                return mySym.equals(symbol);
```

Comment:
Note that `ClassEntry::asSymbol()` also caches array `ClassDesc`s in `ref1.typeSym` so that the symbol cache can be shared with `FieldRefEntry` and `ConstantDynamicEntry`.

--------------------------------------------------------------------------------

I suggest adding the following check:
```java
if (ref1.typeSym instanceof ClassDesc ref1Sym && ref1Sym.isArray()) {
	return ref1Sym.equals(symbol);
}
```

Reply:
I think adding this on the sym present hot path might be too costly. A better way is to pull in the sym before `ref1.equalsString` and push the compatible symbol back to utf8 if we have `equals`.

Reply:
Done.